sql/core/src/main/scala/org/apache/spark/sql/execution/python/VectorizedPythonRunner.scala

```diff
@@ -0,0 +1,329 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.python
+
+import java.io.{BufferedInputStream, BufferedOutputStream, DataInputStream, DataOutputStream}
+import java.net.Socket
+import java.nio.charset.StandardCharsets
+
+import scala.collection.JavaConverters._
+
+import org.apache.arrow.vector.VectorSchemaRoot
+import org.apache.arrow.vector.stream.{ArrowStreamReader, ArrowStreamWriter}
+
+import org.apache.spark.{SparkEnv, SparkFiles, TaskContext}
+import org.apache.spark.api.python.{ChainedPythonFunctions, PythonEvalType, PythonException, PythonRDD, SpecialLengths}
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.execution.arrow.{ArrowUtils, ArrowWriter}
+import org.apache.spark.sql.execution.vectorized.{ArrowColumnVector, ColumnarBatch, ColumnVector}
+import org.apache.spark.sql.types._
+import org.apache.spark.util.Utils
+
+/**
+ * Similar to `PythonRunner`, but exchange data with Python worker via columnar format.
+ */
+class VectorizedPythonRunner(
+    funcs: Seq[ChainedPythonFunctions],
+    batchSize: Int,
+    bufferSize: Int,
+    reuse_worker: Boolean,
+    argOffsets: Array[Array[Int]]) extends Logging {
+
+  require(funcs.length == argOffsets.length, "argOffsets should have the same length as funcs")
+
+  // All the Python functions should have the same exec, version and envvars.
+  private val envVars = funcs.head.funcs.head.envVars
+  private val pythonExec = funcs.head.funcs.head.pythonExec
+  private val pythonVer = funcs.head.funcs.head.pythonVer
+
+  // TODO: support accumulator in multiple UDF
+  private val accumulator = funcs.head.funcs.head.accumulator
+
+  // todo: return column batch?
+  def compute(
```

Comment:
This classes duplicates quite a bit of logic of PythonRDD. I think the only difference is how they serialize/deserialize data (non-arrow vs arrow).  @ueshin @BryanCutler what's your thought on refactoring this and PythonRDD?

Reply:
I agree with you that we should refactor `PythonRunner`s, but I'm not sure what you mean by refactoring PythonRDD? I think it's already simple enough.

Reply:
Yes I meant `PythonRunner` in `PythonRDD.scala`

Reply:
Yes, it is a lot of duplicated code from `PythonRunner` that could be refactored.  I'm guessing you did not use the existing code because of the Arrow stream format?  While I would love to start using that in Spark, I think it would be better to do this at a later time when the required code could be refactored and the Arrow stream format could replace where we currently use the file format.

Also, the good part about using the iterator based file format is each iteration can allow Python to communicate back an error code and exit gracefully.  In my own tests with the streaming format if an error occurred after the stream had started, Spark could lock up in a waiting state.  These are the reasons I did not use the streaming format in my implementation.  Would this `VectorizedPythonRunner` be able to handle these types of errors?

Reply:
@icexelloss Ah, I see, thanks! I still agree with refactoring `PythonRunner`.
@BryanCutler As for the error, do you mean the case like [test_vectorized_udf_exception](https://github.com/apache/spark/pull/19147/files#diff-7c2fe8530271c0635fb99f7b49e0c4a4R3250)? If not, could you please let me know the case and think about it?

Reply:
I was referring to the protocol between Scala and Python that is changed here and could act differently under some circumstances.  Here is the behavior of the `PythonRunner` protocol and `VectorizedPythonRunner` protocol that you introduce here:

**PythonRunner**
Data blocks are framed by a special length integer.  Scala reads each data block one at a time and checks the length code.  If the code is a `PythonException`, the error is read from Python and a `SparkException` is thrown with that being the cause.

**VectorizedPythonRunner**
A data stream is opened in Scala with `ArrowStreamReader` and batches are transferred until `ArrowStreamReader` returns False indicating there is no more data.  Only at this point are the special length codes checked to handle an error from Python.

This behavior change would probably only cause problems if things are not working normally.  For example, what would happen if `pyarrow` was not installed on an executor?  With `PythonRunner` the ImportError would cause a `PythonException` to be transferred and thrown in Scala.  In `VectorizedPythonRunner` I believe the `ArrowStreamReader` would try to read the special length code and then fail somewhere internally to Arrow, not showing the ImportError.

My point was that this type of behavior change should probably be implemented in a separate JIRA where we could make sure to handle all of these cases.